### PR TITLE
WIP backend: add local backend

### LINKED ISSUE #290: Implement data and index kv sort in lightning
## Feature Request

**Is your feature request related to a problem? Please describe:**

To remove importer component. we need to split & scatter region in lightning before generate sst, so the sortkv need to confirm first.

**Describe the feature you'd like:**

1.  add new abstract backend(local)
2. writeRows to local and sort (maybe use goleveldb or something)

**Describe alternatives you've considered:**

1. data kvs can sort partly but index kvs need sort globally

**Teachability, Documentation, Adoption, Optimization:**

### LINKED ISSUE #291: Implement split & scatter in lightning
## Feature Request

**Is your feature request related to a problem? Please describe:**

after #290, we need split & scatter region before send kvs to TiKV peers.

**Describe the feature you'd like:**

port split & scatter from current importer.

**Describe alternatives you've considered:**

maybe we can abstract split & scatter logic, both BR and Lightning can reuse this  

**Teachability, Documentation, Adoption, Optimization:**

### LINKED ISSUE #300: Lightning add local backend
This `local` backend allow user to use `Lightning` independent without `Importer`.
After https://github.com/tikv/tikv/pull/7459 finished, `Lightning` has the ability to write to TiKV alone.

we have follow things to do:
- [x] add local backend for lightning (#326)
- [x] optimize local backend performance (#334, #338)
- [x] add test plan for local backend (https://github.com/pingcap/automated-tests/pull/412)
- [x] update related docs

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve? 
try to resolve #291, resolve #290, resolve #300.
This PR move sort kv and split & scatter from importer into lightning, so that we can remove importer component.
something need to do later:
1. add more general retry logic (not leader error, epoch not match)
2. optimize performance
3. add more unit tests / integration tests

### What is changed and how it works?
1.  add new backend( local which means use local file to sort)
2. transform writeRows write kvs to local file(levelDB)
3. read local file into ranges by regionSplitSize(96M default)
4. batch split & scatter region by ranges
5. concurrent write range to tikv, to generate sst file

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
1. deploy one TiKV/TiDB/PD cluster.
2. use mydumper to dump 20-30 different schemas, the biggest table has 200M data
3. use lightning local backend to import data

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -9,22 +9,28 @@ require (
...
 	github.com/shurcooL/httpgzip v0.0.0-20190720172056-320755c1c1b0
+	github.com/syndtr/goleveldb v1.0.1-0.20190625010220-02440ea7a285
 	go.uber.org/zap v1.14.1
 	golang.org/x/sync v0.0.0-20200317015054-43a5402ce75a
 	golang.org/x/text v0.3.2
 	google.golang.org/grpc v1.25.1
 	modernc.org/mathutil v1.0.0
 )
+
+replace github.com/pingcap/kvproto => github.com/3pointer/kvproto v0.0.0-20200415111719-a032a9dfbfcb
```
Comment (3pointer) [inline go.mod:36]: remove this after https://github.com/pingcap/kvproto/pull/602 merged

File: lightning/config/config.go [hunk 2]
```diff
@@ -150,10 +153,12 @@ type MydumperRuntime struct {
...
-	Addr        string `toml:"addr" json:"addr"`
-	Backend     string `toml:"backend" json:"backend"`
-	OnDuplicate string `toml:"on-duplicate" json:"on-duplicate"`
-	MaxKVPairs  int    `toml:"max-kv-pairs" json:"max-kv-pairs"`
+	Addr            string `toml:"addr" json:"addr"`
+	Backend         string `toml:"backend" json:"backend"`
+	OnDuplicate     string `toml:"on-duplicate" json:"on-duplicate"`
+	MaxKVPairs      int    `toml:"max-kv-pairs" json:"max-kv-pairs"`
+	RegionSplitSize int64  `toml:"region-split-size" json:"region-split-size"`
+	FilePath        string `toml:"file-path" json:"file-path"`
```
Comment (kennytm) [inline lightning/config/config.go:161]: please use a more specific config name

File: lightning/common/security.go [hunk 3]
```diff
@@ -141,3 +148,15 @@ func (tc *TLS) WrapListener(l net.Listener) net.Listener {
...
+
+func (tc *TLS) ToPDSecurityOption() pd.SecurityOption {
+	return pd.SecurityOption{
+		CAPath:   tc.caPath,
+		CertPath: tc.certPath,
+		KeyPath:  tc.keyPath,
+	}
+}
+
+func (tc *TLS) TransToTlsConfig() (*tls.Config, error) {
```
Comment (kennytm) [inline lightning/common/security.go:160]: both us and BR should probably change to use https://github.com/pingcap/tidb-tools/blob/master/pkg/utils/security.go, so we can get rid of this function.

File: lightning/backend/local.go [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+	end    []byte
+	length int
+}
+
+type localFile struct {
+	ts        uint64
+	db        *leveldb.DB
+	meta      sst.SSTMeta
+	length    int
+	totalSize int
```
Comment (kennytm) [inline lightning/backend/local.go:55]:
```suggestion
	length    int64
	totalSize int64
```

File: lightning/backend/local.go [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+	}
+}
+
+func (local *local) RetryImportDelay() time.Duration {
+	return defaultRetryBackoffTime
+}
+
+func (local *local) MaxChunkSize() int {
+	// 31 MB. hardcoded by importer, so do we
+	return 31 << 10
```
Comment (kennytm) [inline lightning/backend/local.go:139]: importer is irrelevant in this backend 🙃

File: lightning/backend/local.go [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+func nextKey(key []byte) []byte {
+	if len(key) == 0 {
+		return []byte{}
+	}
+	res := make([]byte, 0, len(key))
+	s, e := nextKeyNoAlloc(key)
+	res = append(append(res, s...), e)
+	return res
+}
```
Comment (kennytm) [inline lightning/backend/local.go:488]: Merge these two functions.

File: lightning/backend/local.go [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+
+type grpcClis struct {
+	mu   sync.Mutex
+	clis map[uint64]sst.ImportSSTClient
+}
+
+type local struct {
+	mu sync.Mutex
+
+	engines  map[uuid.UUID]localFile
```
Comment (kennytm) [inline lightning/backend/local.go:74]: can we use `sync.Map` for these?

File: lightning/backend/local.go [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+	meta *sst.SSTMeta,
+	ts uint64,
+	region *split.RegionInfo,
+	mutations []*sst.Mutation) (metas []*sst.SSTMeta, err error) {
+	leader := region.Leader
+	if leader == nil {
+		leader = region.Region.GetPeers()[0]
+	}
+
+	cli, err := local.getImportClient(leader)
```
Comment (kennytm) [inline lightning/backend/local.go:184]: the KV pairs should be written to every peer in the region, not just the leader.

**Comment by 3pointer** [comment]
close due to https://github.com/pingcap/tidb-lightning/pull/326

**Pull request closed without merging**